### ISSUE TITLE
next() not raising StopIteration in tf.python.keras.preprocessing.image.DirectoryIterator

### ISSUE THREAD

**Issue by stephengmatthews** [tensorflow/tensorflow#44585]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code: no
- OS Platform and Distribution: Linux Ubuntu 18.04.1
- Mobile device: n/a
- TensorFlow installed from: binary
- TensorFlow version: v2.3.0-54-gfcc4b966f1 2.3.1
- Python version: 3.8.6
- Bazel version: n/a
- GCC/Compiler version: n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Describe the current behavior**
I would like to iterate over image files in a directory using `flow_from_directory()` but calling `next()` does not raise `StopIteration` at the end of the data. Instead, it produces an infinite loop. Is this expected the behavior?

This relates to issues [45](https://github.com/keras-team/keras-preprocessing/issues/45), [226](https://github.com/keras-team/keras-preprocessing/issues/226), and [269](https://github.com/keras-team/keras-preprocessing/issues/269) in the keras preprocessing repo. I understand development has moved to tensorflow, hence posting it here.

**Describe the expected behavior**
I expected `StopIteration` to be raised.

**Standalone code to reproduce the issue**
The code below was adapted from [here](https://github.com/keras-team/keras-preprocessing/issues/226#issuecomment-516648880).

```
from tensorflow.keras.preprocessing.image import ImageDataGenerator, save_img
import numpy as np
import os

batch_size = 3
n_images = 5

directory = '/tmp/test'
classes = ['dog', 'cat']

for i in range(n_images):
    label = np.random.choice(classes)
    os.makedirs(os.path.join(directory, label), exist_ok=True)
    filename = os.path.join(directory, label, str(i) + '.jpg')
    save_img(filename, np.ones((64, 64, 3)))

gen = ImageDataGenerator().flow_from_directory(
    '/tmp/test',
    class_mode=None,
    batch_size=batch_size
)

for _ in range(gen.__len__()):
    imgs = next(gen)
    print(imgs.shape)

gen.reset()
while True:
    imgs = next(gen)
    print(imgs.shape)
```

**Other info / logs**
n/a

**Comment by Saduf2019** [comment]
@stephengmatthews 
I ran the code shared, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/0a2d008b43a7d31de6efb51fb6cea04d/untitled458.ipynb) please confirm, this doesn't seem like a bug at tf end.

**Comment by stephengmatthews** [comment]
Thank you. My assumption is that `next()` handles the end of iteration, similar to Python's built-in [`next()`](https://docs.python.org/3.8/library/functions.html#next) and `tf.data.Iterator`'s [`get_next()`](https://www.tensorflow.org/api_docs/python/tf/data/Iterator#get_next).

Could the use case for `next()` looping indefinitely be clarified, please? This [training workflow guide](https://www.tensorflow.org/guide/data#processing_multiple_epochs) suggests that the use case is similar to using `Dataset.repeat()` "to iterate over a dataset in multiple epochs".

[Older versions](https://faroit.com/keras-docs/2.1.1/preprocessing/image/) of the keras API state that "data will be looped over (in batches) indefinitely". The tensorflow API also mentions this in an [example (see penultimate line)](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/ImageDataGenerator#examples_2). I wonder whether it could be clarified explicitly in the documentation (rather than buried as a comment in a code snippet). For me at least, [older versions](https://faroit.com/keras-docs/2.1.1/preprocessing/image/) of the keras API were much easier to understand that the loop is indefinite without some digging around.

What are your thoughts on this?

**Comment by gowthamkpr** [comment]
@stephengmatthews 
`Deprecated`: [tf.keras.preprocessing.image](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image) APIs do not operate on tensors and are not recommended for new code. Prefer loading data with [tf.keras.utils.image_dataset_from_directory](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory), and then transforming the output [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) with preprocessing layers. For more information, see the tutorials for [loading images](https://www.tensorflow.org/tutorials/load_data/images) and [augmenting images](https://www.tensorflow.org/tutorials/images/data_augmentation), as well as the [preprocessing layer guide](https://www.tensorflow.org/guide/keras/preprocessing_layers).


Please use the newer apis where this issue doesn't exist. Thanks!!!

**Comment by YunusBulut** [comment]
This is still an issue. I'm using tensorflow==2.13.0rc0, tensorflow-macos==2.13.0rc0 and the loop funs forever as long as you set shuffle parameter to False.
